nodes/address.go

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nodes
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"sort"
+)
+
+// GuessNodeExternalAddress tries to guess external address of a node
+func (n *NodesClient) GuessNodeExternalAddress(node *corev1.Node) *corev1.NodeAddress {
+	sorted := prioritizeAddresses(node.Status.Addresses)
+	first := sorted[0]
+	priority := addressToPriority(first)
+	if priority >= 2 {
+		n.logger.Warnf("Chosen address is probably an internal type: %s, "+
+			"and might be unaccessible", first.Type)
+	}
+	return first
+}
+
+type nodeAddresses []*corev1.NodeAddress
+
+type byAddressType struct {
+	nodeAddresses
+}
+
+func (s byAddressType) Len() int {
+	return len(s.nodeAddresses)
+}
+
+func (s byAddressType) Swap(i, j int) {
+	s.nodeAddresses[i], s.nodeAddresses[j] = s.nodeAddresses[j], s.nodeAddresses[i]
+}
+
+func (s byAddressType) Less(i, j int) bool {
+	return addressToPriority(s.nodeAddresses[i]) < addressToPriority(s.nodeAddresses[j])
+}
+
+var addressTypeToPriorities = map[corev1.NodeAddressType]int{
```

Comment:
How you would enum here? I need to convert address type to priority.

Reply:
You can define your enums with integer types (see go enums)
then you just do `priority > nodeip`.

Reply:
I think it's impossible to do as you asked. `corev1.NodeAddressType` is string type. We can create Golang enum of string type, but it will be a bunch of constants, without ordering.

```golang
const (
  NodeExternalDNS = "ExternalDNS"
  NodeExternalIP  = "ExternalIP"
  # [..]
)
```

This map "converts" `corev1.NodeAddressType` to int value that can be sorted.